File: src/aistudio_client/makersuite/tools.py
"""Toolهای Vertex را به ساختار positional MakerSuite تبدیل می‌کند."""

from typing import Any

from .schema import encode_schema


def encode_tools(tools: list[Any] | None) -> list | None:
    if tools is None:
        return None
    return [_encode_tool(tool) for tool in tools]


def _encode_tool(tool: Any) -> list:
    if not isinstance(tool, dict):
        raise ValueError("Tool must be an object")
    supported = {"codeExecution", "functionDeclarations"}
    unknown = set(tool) - supported
    if unknown:
        names = ", ".join(sorted(unknown))
        raise ValueError(f"Unsupported Tool fields: {names}")

    encoded = [None, None]
    if tool.get("codeExecution") is not None:
        encoded[0] = []
    declarations = tool.get("functionDeclarations")
    if declarations is not None:
        encoded[1] = [_function_declaration(item) for item in declarations]
    if all(value is None for value in encoded):
        raise ValueError("Tool must set codeExecution or functionDeclarations")
    return _trim(encoded)


def _function_declaration(value: Any) -> list:
    if not isinstance(value, dict) or not value.get("name"):
        raise ValueError("functionDeclarations[].name is required")
    parameters = value.get("parameters")
    if parameters is None:
        parameters = value.get("parametersJsonSchema")
    response = value.get("response")
    if response is None:
        response = value.get("responseJsonSchema")
    return _trim([
        value["name"],
        value.get("description"),
        encode_schema(parameters) if parameters is not None else None,
        encode_schema(response) if response is not None else None,
    ])


def _trim(values: list) -> list:
    while values and values[-1] is None:
        values.pop()
    return values

File: internal/aistudio/tools.go
package aistudio

import (
	"fmt"
	"sort"
	"strings"
)

func EncodeTools(tools []any) (any, error) {
	if tools == nil {
		return nil, nil
	}
	result := make([]any, 0, len(tools))
	for _, raw := range tools {
		tool, ok := raw.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("Tool must be an object")
		}
		unknown := []string{}
		for name := range tool {
			if name != "codeExecution" && name != "functionDeclarations" {
				unknown = append(unknown, name)
			}
		}
		if len(unknown) > 0 {
			sort.Strings(unknown)
			return nil, fmt.Errorf("Unsupported Tool fields: %s", strings.Join(unknown, ", "))
		}
		encoded := make([]any, 2)
		if tool["codeExecution"] != nil {
			encoded[0] = []any{}
		}
		if declarations := tool["functionDeclarations"]; declarations != nil {
			items, ok := declarations.([]any)
			if !ok {
				return nil, fmt.Errorf("functionDeclarations must be an array")
			}
			values := make([]any, 0, len(items))
			for _, item := range items {
				declaration, err := encodeFunctionDeclaration(item)
				if err != nil {
					return nil, err
				}
				values = append(values, declaration)
			}
			encoded[1] = values
		}
		if encoded[0] == nil && encoded[1] == nil {
			return nil, fmt.Errorf("Tool must set codeExecution or functionDeclarations")
		}
		result = append(result, trim(encoded))
	}
	return result, nil
}

func encodeFunctionDeclaration(value any) ([]any, error) {
	declaration, ok := value.(map[string]any)
	if !ok || declaration["name"] == nil || fmt.Sprint(declaration["name"]) == "" {
		return nil, fmt.Errorf("functionDeclarations[].name is required")
	}
	parameters := declaration["parameters"]
	if parameters == nil {
		parameters = declaration["parametersJsonSchema"]
	}
	response := declaration["response"]
	if response == nil {
		response = declaration["responseJsonSchema"]
	}
	var encodedParameters, encodedResponse any
	var err error
	if parameters != nil {
		encodedParameters, err = EncodeSchema(parameters)
		if err != nil {
			return nil, err
		}
	}
	if response != nil {
		encodedResponse, err = EncodeSchema(response)
		if err != nil {
			return nil, err
		}
	}
	return trim([]any{declaration["name"], declaration["description"], encodedParameters, encodedResponse}), nil
}
